src/main/scala/inox/ast/SymbolOps.scala

```diff
@@ -115,6 +115,44 @@ trait SymbolOps { self: TypeOps =>
     case _ => false
   }
 
+  /** Returns true when the given expression is ground, pure, and does not contain
+   *  choose or quantifiers, and can thus be simplified.
+   */
+  def canBeForced(expr: Expr): Boolean = {
+    isGround(expr) && isPure(expr)(PurityOptions.AssumeChecked) && !containsChooseOrQuantifiers(expr)
+  }
+
+  /** Simplify all the pure and ground sub-expressions of the given expression,
+   * by forcing them using [[inox.evaluators.Evaluator]].
+   *
+   * @see [[canBeForced]]
+   */
+  def simplifyGround(expr: Expr)(implicit sem: symbols.Semantics, ctx: Context): Expr = {
+    val evaluator = sem.getEvaluator
+
+    def err(desc: String, msg: String): Unit = {
+      ctx.reporter.error(s"Forcing failed due to a $desc: $msg")
+    }
+
+    import inox.evaluators.EvaluationResults._
+
+    postMap {
```

Comment:
Using a postMap here is valid but you will be invoking the evaluator many more times than necessary. For the `RecursiveEvaluator`, this probably doesn't matter much but with the `CodeGenEvaluator`, you get a non-trivial setup time (although evaluation is then much faster). I would use a dedicated recursive function that doesn't call itself on the evaluation result.

Reply:
Yes I figured that this is not the optimal way to do such things. Should the transformation be top-down or bottom-up then? Can you sketch (in a few words, no need for code except if it's easier for you to explain it that way), how such a recursive function would look like?  Something like ?

```scala
def rec(e: Expr): Expr = e match {
  case fi @ FunctionInvocation(_, _, args) => ???
  case Let(i, e, b) => ???
  case Operator(es, recons) => ???
}
```

Reply:
Top down I'd say. As soon as you find something that is pure and ground, I'd try to simplify it. If the simplification fails, then recurse into the children. If it succeeds, then you have a value and are done. You can add some more checks if you want to avoid calling the evaluator when it's obvious it will fail but I'd just start with this for now and see whether you get a slowdown on existing benchmarks. I don't think we have so many ground expressions so it probably won't slow things down much. You can also avoid calling the simplifier if the expression is already a value.

So the cases in your function would be something like:
```scala
case e if isValue(e) => e
case e if isPure(e) && isGround(e) => // eval
case Operator(es, recons) => recons(es.map(rec))
```